docker/transport/sshconn.py

```diff
@@ -20,33 +23,140 @@
 RecentlyUsedContainer = urllib3._collections.RecentlyUsedContainer
 
 
+def create_paramiko_client(base_url):
+    logging.getLogger("paramiko").setLevel(logging.WARNING)
+    ssh_client = paramiko.SSHClient()
+    base_url = six.moves.urllib_parse.urlparse(base_url)
+    ssh_params = {
+        "hostname": base_url.hostname,
+        "port": base_url.port,
+        "username": base_url.username
+        }
+    ssh_config_file = os.path.expanduser("~/.ssh/config")
+    if os.path.exists(ssh_config_file):
+        conf = paramiko.SSHConfig()
+        with open(ssh_config_file) as f:
+            conf.parse(f)
+        host_config = conf.lookup(base_url.hostname)
+        ssh_conf = host_config
+        if 'proxycommand' in host_config:
+            ssh_params["sock"] = paramiko.ProxyCommand(
+                ssh_conf['proxycommand']
+            )
+        if 'hostname' in host_config:
+            ssh_params['hostname'] = host_config['hostname']
+        if 'identityfile' in host_config:
+            ssh_params['key_filename'] = host_config['identityfile']
+        if base_url.port is None and 'port' in host_config:
+            ssh_params['port'] = ssh_conf['port']
+        if base_url.username is None and 'user' in host_config:
+            ssh_params['username'] = ssh_conf['user']
+
+    ssh_client.load_system_host_keys()
+    ssh_client.set_missing_host_key_policy(paramiko.WarningPolicy())
+    return ssh_client, ssh_params
+
+
+class SSHSocket(socket.socket):
+    def __init__(self, host):
+        super(SSHSocket, self).__init__(
+            socket.AF_INET, socket.SOCK_STREAM)
+        self.host = host
+        self.port = None
+        if ':' in host:
+            self.host, self.port = host.split(':')
+        self.proc = None
+
+    def connect(self, **kwargs):
+        port = '' if not self.port else '-p {}'.format(self.port)
+        args = [
+            'ssh',
+            '-q',
+            self.host,
+            port,
+            'docker system dial-stdio'
+        ]
+        self.proc = subprocess.Popen(
+            ' '.join(args),
+            shell=True,
+            stdout=subprocess.PIPE,
+            stdin=subprocess.PIPE)
+
+    def sendall(self, msg):
+        if not self.proc or self.proc.stdin.closed:
+            raise Exception('SSH subprocess not initiated.'
+                            'connect() must be called first.')
+        self.proc.stdin.write(msg)
+        self.proc.stdin.flush()
+
+    def send(self, msg):
+        self.sendall(msg)
+        return len(msg)
```

Comment:
nit: To avoid having to compute the message length separately, I would call `send` from `sendall` and use the byte count from the `write()` function's return.

```suggestion
        self.send(msg)

    def send(self, msg):
        if not self.proc or self.proc.stdin.closed:
            raise Exception('SSH subprocess not initiated.'
                            'connect() must be called first.')
        written = self.proc.stdin.write(msg)
        self.proc.stdin.flush()
        return written
```

Reply:
Py2.7 request library seems to have a problem if we call `send` from `sendall`. Moved the logic intro a private method to be called in these two.